tests/test_image.h

```diff
@@ -0,0 +1,283 @@
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef TEST_IMAGE_H
+#define TEST_IMAGE_H
+
+#include "core/io/file_access_pack.h"
+#include "core/io/image.h"
+
+#include "thirdparty/doctest/doctest.h"
+
+namespace TestImage {
+
+TEST_CASE("[Image] Ways of creating images") {
+	Ref<Image> image = memnew(Image(8, 4, false, Image::FORMAT_RGBA8));
+	CHECK_MESSAGE(
+			!image->is_empty(),
+			"An image created with specified size and format should not be empty at first.");
+	CHECK_MESSAGE(
+			image->is_invisible(),
+			"A newly created image should be invisible.");
+	CHECK_MESSAGE(
+			!image->is_compressed(),
+			"A newly created image should not be compressed.");
+	CHECK_MESSAGE(
+			!image->has_mipmaps(),
+			"has_mipmaps() return the expected value.");
+
+	Ref<Image> image_copy = memnew(Image());
+	CHECK_MESSAGE(
+			image_copy->is_empty(),
+			"An image created without any specified size and format be empty at first.");
+	image_copy->copy_internals_from(image);
+
+	CHECK_MESSAGE(
+			image->get_data() == image_copy->get_data(),
+			"Duplicated images should have the same data.");
+
+	PackedByteArray image_data = image->get_data();
+	Ref<Image> image_from_data = memnew(Image(8, 4, false, Image::FORMAT_RGBA8, image_data));
+	CHECK_MESSAGE(
+			image->get_data() == image_from_data->get_data(),
+			"An image created from data of another image should have the same data of the original image.");
+}
+
+TEST_CASE("[Image] Saving and loading") {
+	Ref<Image> image = memnew(Image(4, 4, false, Image::FORMAT_RGBA8));
+	const String save_path_png = OS::get_singleton()->get_cache_path().plus_file("image.png");
+	const String save_path_exr = OS::get_singleton()->get_cache_path().plus_file("image.exr");
+
+	// Save PNG
+	Error err;
+	err = image->save_png(save_path_png);
+	CHECK_MESSAGE(
+			err == OK,
+			"The image should be saved successfully as a .png file.");
+
+	// Save EXR
+	err = image->save_exr(save_path_exr, false);
+	CHECK_MESSAGE(
+			err == OK,
+			"The image should be saved successfully as an .exr file.");
+
+	// Load using load()
+	Ref<Image> image_load = memnew(Image());
+	err = image_load->load(save_path_png);
+	CHECK_MESSAGE(
+			err == OK,
+			"The image should load successfully using load().");
+	CHECK_MESSAGE(
+			image->get_data() == image_load->get_data(),
+			"The loaded image should have the same data as the one that got saved.");
+
+	const String base_dir = OS::get_singleton()->get_executable_path().get_base_dir();
+
+	// Load BMP
+	const String input_bmp_path = base_dir.plus_file("../tests/test_image/icon.bmp");
```

Comment:
```suggestion
	// Load BMP
	const String input_bmp_path = TestUtils::get_data_path("icon.bmp");
```

To use `TestUtils::get_data_path`, you'll also need to include `#include "test_utils.h"`

But for this to work properly, you'll also need to move image files to `tests/data` directory which is created exactly for this. I think you can create a subfolder, like `tests/data/images` if you don't want to "clutter" the base test data directory. You'll then use relative path `TestUtils::get_data_path("images/icon.bmp")`.

You can also immediately open the BMP file without creating intermediate `input_bmp_path`:

```c++
FileAccessRef f_bmp = FileAccess::open(TestUtils::get_data_path("images/icon.bmp"), FileAccess::READ, &err);
```

The same applies to other paths.

Reply:
Done. I didn't know about TestUtils, this is a much better way, thanks!